Python/ceval.c

```diff
@@ -439,35 +439,83 @@ int pthread_attr_destroy(pthread_attr_t *a)
 #endif
 
 
+static void
+tstate_set_stack(PyThreadState *tstate,
+                 void *stack_start_addr, size_t stack_size)
+{
+    assert(stack_size > 0);
+
+    _PyThreadStateImpl *ts = (_PyThreadStateImpl *)tstate;
+    ts->c_stack_hard_limit = (uintptr_t)stack_start_addr;
+    ts->c_stack_top = (uintptr_t)stack_start_addr + stack_size;
+
+    uintptr_t soft_limit = ts->c_stack_hard_limit;
+    if (stack_size >= _PyOS_STACK_MARGIN_BYTES) {
+#ifdef _Py_THREAD_SANITIZER
+        // Thread sanitizer crashes if we use a bit more than half the stack.
+        soft_limit += (stack_size / 2);
+#else
+        soft_limit += _PyOS_STACK_MARGIN_BYTES;
+#endif
+    }
+    ts->c_stack_soft_limit = soft_limit;
+
+    // Sanity checks
+    assert(ts->c_stack_hard_limit <= ts->c_stack_soft_limit);
+    assert(ts->c_stack_soft_limit < ts->c_stack_top);
+
+    // Test the stack pointer
+#if !defined(NDEBUG) && !defined(__wasi__)
+    uintptr_t here_addr = _Py_get_machine_stack_pointer();
+    assert(ts->c_stack_soft_limit < here_addr);
+    assert(here_addr < ts->c_stack_top);
+#endif
+}
+
+
+int
+PyUnstable_ThreadState_SetStack(PyThreadState *tstate,
+                                void *stack_start_addr, size_t stack_size)
+{
+    if (stack_size == 0) {
+        PyErr_SetString(PyExc_ValueError, "stack_size must be greater than 0");
+        return -1;
+    }
+
+    tstate_set_stack(tstate, stack_start_addr, stack_size);
+    return 0;
+}
+
+
 void
-_Py_InitializeRecursionLimits(PyThreadState *tstate)
+PyUnstable_ThreadState_ResetStack(PyThreadState *tstate)
```

Comment:
This also won't work. We need to record the top of the stack when initially getting the stack limits, and use that instead of `_Py_get_machine_stack_pointer()` otherwise, we would continually lower the stack limits, so that they become invalid.

Reply:
I fixed PyUnstable_ThreadState_ResetStack() to save the initial stack start address and stack size.